src/Chainweb/Rosetta/Util.hs

```diff
@@ -0,0 +1,686 @@
+  { _genesisMatchFunction :: Either RosettaFailure a }
+newtype RemediationMatchFunction a = RemediationMatchFunction
+  { _remediationMatchFunction :: Either RosettaFailure a }
+newtype NonGenesisMatchFunction a = NonGenesisMatchFunction
+  { _nonGenesisMatchFunction :: Either RosettaFailure a }
+
+-- | Retrieve the coin contract logs for transaction(s) in a block.
+matchLogs
+    :: BlockHeader
+    -> GenesisMatchFunction a
+    -> RemediationMatchFunction a
+    -> NonGenesisMatchFunction a
+    -> Either RosettaFailure a
+matchLogs bh g r f
+  | bheight == 0 = _genesisMatchFunction g
+  | coinV2Upgrade v cid bheight = _remediationMatchFunction r
+  | otherwise = _nonGenesisMatchFunction f
+  where
+    bheight = _blockHeight bh
+    cid = _blockChainId bh
+    v = _blockChainwebVersion bh
+
+
+---------------------
+-- Genesis Helpers --
+---------------------
+
+-- |
+-- | Genesis transactions do not have coinbase or gas payments.
+-- |
+getGenesisLog
+    :: Map TxId [AccountLog]
+    -> CommandResult Hash
+    -> Transaction
+getGenesisLog logs cr =
+  case (_crTxId cr) of
+    Just tid -> case (M.lookup tid logs) of
+      Just l -> rosettaTransaction cr $ makeOps tid l
+      Nothing -> rosettaTransaction cr []  -- not a coin contract tx
+    Nothing -> rosettaTransaction cr [] -- all genesis tx should have a txid
+  where
+    makeOps tid l = indexedOperations $
+      map (operation Successful TransferOrCreateAcct tid) l
+
+
+-- |
+-- | Matches all genesis transactions to their coin contract logs.
+-- |
+genesisTransactions
+    :: Map TxId [AccountLog]
+    -> V.Vector (CommandResult Hash)
+    -> GenesisMatchFunction [Transaction]
+genesisTransactions logs txs =
+  GenesisMatchFunction $
+  pure $ V.toList $ V.map (getGenesisLog logs) txs
+
+
+-- |
+-- | Matches a single genesis transaction to its coin contract logs.
+-- |
+genesisTransaction
+    :: Map TxId [AccountLog]
+    -> V.Vector (CommandResult Hash)
+    -> RequestKey
+    -- ^ target tx
+    -> GenesisMatchFunction Transaction
+genesisTransaction logs rest target = GenesisMatchFunction $ do
+  cr <- note RosettaTxIdNotFound $
+        V.find (\c -> (_crReqKey c) == target) rest
+  pure $ getGenesisLog logs cr
+
+
+------------------------
+-- Coinbase Helpers --
+------------------------
+
+-- | Matches the first coin contract logs to the coinbase tx
+nonGenesisCoinbase
+    :: (TxId, [AccountLog])
+    -> (a -> Maybe TxId)
+    -> (a -> [Operation] -> b)
+    -> a
+    -> Either String b
+nonGenesisCoinbase (tid,l) getTxId f cr = do
+  expectedTid <- note "No Coinbase TxId found" (getTxId cr)
+  if (expectedTid == tid)
+    then let ops = indexedOperations $
+                   map (operation Successful CoinbaseReward tid) l
+         in pure $ (f cr ops)
+    else Left "First log's txId does not match coinbase tx's TxId"
+
+
+------------------------
+-- NonGenesis Helpers --
+------------------------
+
+-- | Algorithm:
```

Comment:
Start this comment with a haddocks e.g. "accumulates a triple of transactions representing a tx and it's bracketing gas logs" or whatever.

Reply:
Added